rust/pact_verifier/src/pact_broker.rs

```diff
@@ -497,11 +508,140 @@ pub async fn fetch_pacts_from_broker(
+      })?;
+
+    // Construct the Pacts for verification payload
+    let pacts_for_verification = PactsForVerificationRequest {
+      provider_version_tags: provider_tags,
+      include_wip_pacts_since: include_wip_pacts_since,
+      consumer_version_selectors: consumer_version_selectors,
+      include_pending_status: pending,
+    };
+    let request_body = serde_json::to_string(&pacts_for_verification).unwrap();
+
+    // Post the verification request
+    let response = match hal_client.find_link("self") {
+      Ok(link) => {
+        // TODO: make this an async stream to make error handling less shit
+        match hal_client.clone().post_json(hal_client.clone().parse_link_url(&link, &hashmap!{})?, request_body).await {
+          Ok(res) => Some(res),
+          Err(err) => {
+            debug!("error Response for pacts for verification {:?} ", err);
+            return Err(err)
+          }
+        }
+      },
+      Err(e) => return Err(e)
+    };
+
+    // Find all of the Pact links
+    let pact_links = match response {
+      Some(v) => {
+        let pfv: PactsForVerificationResponse = serde_json::from_value(v).unwrap();
+
+        let links: Result<Vec<(Link, PactVerificationContext)>, PactBrokerError> = pfv.embedded.pacts.iter().map(| p| {
+          match p.links.get("self") {
+            Some(l) => Ok((l.clone(), PactVerificationContext{
+              short_description: p.short_description.clone(),
+              verification_properties: PactVerificationProperties {
+                pending: p.verification_properties.pending,
+                notices: p.verification_properties.notices.clone(),
+              }
+            })),
+            None => Err(
+              PactBrokerError::LinkError(
+                format!(
+                  "Expected a HAL+JSON response from the pact broker, but got a link with no HREF. URL: '{}', PATH: '{:?}'",
+                  &hal_client.url,
+                  &p.links,
+                )
+              )
+            )
+          }
+        }).collect();
+
+        links
+      },
+      None => Err(PactBrokerError::NotFound(format!("No pacts were found for this provider")))
+    }?;
+
+    let results: Vec<_> = futures::stream::iter(pact_links)
+        // TODO: this first branch is a dupe of the above, I think
+        .map(|(ref pact_link, ref context)| {
+          match pact_link.href {
+            Some(_) => Ok((hal_client.clone(), pact_link.clone(), context.clone())),
+            None => Err(
+              PactBrokerError::LinkError(
+                format!(
+                  "Expected a HAL+JSON response from the pact broker, but got a link with no HREF. URL: '{}', LINK: '{:?}'",
+                  &hal_client.url,
+                  pact_link
+                )
+              )
+            )
+          }
+        })
+        .and_then(|(hal_client, pact_link, context)| async {
+          let pact_json = hal_client.fetch_url(
+            &pact_link.clone(),
```

Comment:
Because the fetch_url takes references, you might be able to not need the clones with a reference, but just add a ref to the parameters to the closure. But I say might, because the dependency checker is mean.

Reply:
It turns out the checker is rather mean.